astropy/io/fits/fitsrec.py

```diff
@@ -823,7 +823,7 @@ def _convert_ascii(self, column, field):
         """
 
         format = column.format
-        recformat = ASCII2NUMPY[format[0]]
+        recformat = getattr(column, 'dtype', ASCII2NUMPY[format[0]])
```

Comment:
Not sure about this, I think I see how the old code could be an issue, but I wonder why this was not using `column.recformat` directly ?

Reply:
Should better document my changes to myself, it's already too long ago...
The core of the issue was, I think, that `_convert_ascii` might be called on a column that does not yet have a `recformat` set (I would have thought that `Table.read()` would already read in the table with `data` as `np.array` of the appropriate `dtype`, but it turned out it is initially stored as `str` columns, which are only converted with `_convert_ascii` “on demand”).
In any case `getattr(column, 'recformat')` still reads the example file from #9899 incorrectly as `int32`, although it passes all the tests, so I guess there is another test to be added for that...

Reply:
Hum, this is why io.fits code is hard to change :D 
I also forgot the details but in the past I tried to replace this line with `format.recformat` (https://github.com/astropy/astropy/pull/5362/commits/0fc487e8a8b43ed1cfb0f2e4a53566021042171d) and reverted this in the commit after because we chose to read all float values as double (#5362, #5353).
Maybe we should just do the same for integers and always read as int64 ? (cc @MSeifert04 if you have a better memory than me ;)).

Reply:
Ah, `column.recformat` did not work, because the `_AsciiColumnFormat` method is just a wrapper around `_convert_ascii_format` and will thus determine `recformat` anew from `format` (maybe it should first check if `self.recformat` or `self.format.recformat` already exist?).

Reply:
yes sorry that's `format.recformat`, but if we change the default value that does not matter anymore.

> So I think changing results by making int64 always the default should be deferred to a PR for a later milestone. In terms of memory efficiency it would probably be preferable to stick with the current tailored formats (although you could say if memory usage becomes an issue, ascii formats are hardly a sensible choice anyway).

Memory usage should not be an issue, in particular for ascii files. I prefer that we fix the bug and produce valid results, so I think it is fine with the next bugfix release.

Reply:
Keeping with int types tailored to the column makes sense.  For example, sometimes there are "boolean" columns that only contain 0 or 1--it makes to sense to waste 8 bytes per row on such columns, when an 8-bit int can be used instead.